src/psyclone/transformations.py

```diff
@@ -2443,20 +2446,55 @@ def validate(self, node, options=None):
                 raise TransformationError(
                     f"Failed to create PSyIR for kernel '{node.name}'. "
                     f"Cannot transform such a kernel.") from error
-
-            # Check that the kernel does not access any data or routines via a
-            # module 'use' statement
-            imported_variables = kernel_schedule.symbol_table.imported_symbols
-            if imported_variables:
+            k_or_r = "Kernel"
+        else:
+            # Supplied node is a PSyIR Routine which *is* a Schedule.
+            kernel_schedule = node
+            k_or_r = "routine"
+
+        # Check that the routine does not access any data that is imported via
+        # a 'use' statement.
+        refs = kernel_schedule.walk(Reference)
+        for ref in refs:
+            if ref.symbol.is_import:
+                # resolve_deferred does nothing if the Symbol type is known.
+                try:
+                    ref.symbol.resolve_deferred()
+                except SymbolError:
+                    # TODO #11 - log that we failed to resolve this Symbol.
+                    pass
```

Comment:
This is just a question/comment, not for this PR, but some transformations do not resolve deferred and instead produce validation errors, and then the script writer can resolve before calling the transformation (e.g. in nemo); another couple of places do the resolve when needed (like here), and in some cases it would be useful to have them earlier than the script, like during the reader (e.g. where cannonicalisation).

We should rethink/clarify what is the best practice here? And maybe even resolve from all containers that are listed on the "-I" files from the very beginning, so we always start with all information we have.

Reply:
Yes, I'm beginning to think that would be best. Feels like a very similar conversation to the one in #2296.